### PR TITLE
go_1_12: init at go 1.12

### PR DESCRIPTION
**Pull request by kalbasit**
###### Motivation for this change

The [build cache](https://golang.org/cmd/go/#hdr-Build_and_test_caching) is now required as a step toward eliminating $GOPATH/pkg. Setting the environment variable GOCACHE=off will cause go commands that write to the cache to fail. I've updated all the packages that I could find and are using go without using `buildGoPackage`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS (sandbox off)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Tested with

```console
λ nix run -f '<nixpkgs>' nur.repos.kalbasit.nix-verify -c nix-verify -I nixpkgs=`pwd` \
        -A bazel-watcher \
        -A boringssl \
        -A cadvisor \
        -A cide \
        -A cni-plugins \
        -A deepin.dbus-factory \
        -A deepin.dde-api \
        -A deepin.dde-daemon \
        -A deepin.deepin-desktop-base \
        -A deepin.deepin-desktop-schemas \
        -A deepin.deepin-metacity \
        -A deepin.deepin-wallpapers \
        -A deepin.deepin-wm \
        -A deepin.go-dbus-generator \
        -A deepin.go-gir-generator \
        -A docker \
        -A docker-gc \
        -A gitlab-shell \
        -A gitlab-workhorse \
        -A go \
        -A go_1_11 \
        -A gometalinter \
        -A kubectl \
        -A kubectx \
        -A kubernetes \
        -A pipework \
        -A prometheus \
        -A prometheus-alertmanager \
        -A prometheus-pushgateway \
        -A prometheus_2 \
        -A python37Packages.jupyter-repo2docker \
        -A rkt \
        -A singularity \
        -A tmsu \
        -A vault \
        -A vgo2nix \
        -A vimPlugins.vim-go
```

### CODE REVIEW

**Comment by kalbasit** [comment]
@GrahamcOfBorg bazel-watcher boringssl cadvisor cide cni-plugins deepin.dbus-factory deepin.dde-api deepin.dde-daemon deepin.deepin-desktop-base deepin.deepin-desktop-schemas deepin.deepin-metacity deepin.deepin-wallpapers deepin.deepin-wm deepin.go-dbus-generator deepin.go-gir-generator docker docker-gc gitlab-shell gitlab-workhorse go go_1_11 gometalinter kubectl kubectx kubernetes pipework prometheus prometheus-alertmanager prometheus-pushgateway prometheus_2 python37Packages.jupyter-repo2docker rkt singularity vault vgo2nix vimPlugins.vim-go

**Comment by kalbasit** [comment]
@GrahamcOfBorg build bazel-watcher boringssl cadvisor cide cni-plugins deepin.dbus-factory deepin.dde-api deepin.dde-daemon deepin.deepin-desktop-base deepin.deepin-desktop-schemas deepin.deepin-metacity deepin.deepin-wallpapers deepin.deepin-wm deepin.go-dbus-generator deepin.go-gir-generator docker docker-gc gitlab-shell gitlab-workhorse go go_1_11 gometalinter kubectl kubectx kubernetes pipework prometheus prometheus-alertmanager prometheus-pushgateway prometheus_2 python37Packages.jupyter-repo2docker rkt singularity vault vgo2nix vimPlugins.vim-go

**Comment by Mic92** [comment]
The checksum error on darwin should be fixed though:

@GrahamcOfBorg build go_1_12

**Comment by kalbasit** [comment]
@Mic92 I suspect that it is, as well. I tested the entire PR on Darwin myself, but unlike ofborg I’m running APPFS.

File: pkgs/development/compilers/go/1.12.nix [hunk 1]
```diff
@@ -31,11 +31,9 @@ stdenv.mkDerivation rec {
   name = "go-${version}";
   version = "1.12";
 
-  src = fetchFromGitHub {
-    owner = "golang";
-    repo = "go";
-    rev = "go${version}";
-    sha256 = "08j7ghcy5cs5p4sw5rqi57lzg52lrix8xbxn87y7y9sv6s3wx44n";
+  src = fetchurl {
+    url = "https://github.com/golang/go/archive/go${version}.tar.gz";
```
Comment (Mic92) [inline pkgs/development/compilers/go/1.12.nix]: I would say just use `https://dl.google.com/go/go1.12.src.tar.gz` instead. It is the official source anyway
and release archives should not be regenerated.

Comment (kalbasit) [reply pkgs/development/compilers/go/1.12.nix]: Good point! I switched it over.

**Comment by Mic92** [comment]
@GrahamcOfBorg build go_1_12

**Comment by Mic92** [comment]
```
@GrahamcOfBorg build prometheus
```

**Comment by Mic92** [comment]
The last darwin build error just look like resource depletion. I think we can ignore this given that you already build on macOS.

**Comment by Mic92** [comment]
This was the build error log of prometheus: https://dl.thalheim.io/RRa5YdOzO7l0jKJ1cV01aA/prometheus.log

**Comment by Mic92** [comment]
Ok. Maybe this was caused by my zfs filesystem or some other random stuff.

File: pkgs/development/compilers/go/1.12.nix [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    ln -s $GOROOT_FINAL/bin $out/bin
+    runHook postInstall
+  '';
+
+  setupHook = ./setup-hook.sh;
+
+  disallowedReferences = [ goBootstrap ];
+
+  meta = with stdenv.lib; {
+    branch = "1.11";
```
Comment (dtzWill) [inline pkgs/development/compilers/go/1.12.nix]: Probably should be `branch = "1.12";`? (This was in an earlier version of this PR, looks like the commit was dropped?)

Comment (kalbasit) [reply pkgs/development/compilers/go/1.12.nix]: oh, shoot, how did this happen? I'll file a subsequent PR to fix it. Thanks!

Comment (kalbasit) [reply pkgs/development/compilers/go/1.12.nix]: @dtzWill I filed https://github.com/NixOS/nixpkgs/pull/56852 to address this. Thank you!

Comment (Mic92) [reply pkgs/development/compilers/go/1.12.nix]: Not too bad though. I am not aware of any usage of this attribute.

**Pull request merged**